File: week01/c/03_structs.c
/*
 * 03_structs.c - Struct definition and usage
 *
 * Structs group related data together into a single type.
 * They are the foundation of data organization in C.
 */

#include <stdio.h>
#include <string.h>

/* Basic struct definition */
struct Point {
    int x;
    int y;
};

/* Struct with different field types */
struct Person {
    char name[50];
    int age;
    double height;
};

/* Using typedef to avoid writing 'struct' everywhere */
typedef struct {
    char text[100];
    int line;
    int column;
} Token;

/* Nested structs */
struct Rectangle {
    struct Point top_left;
    struct Point bottom_right;
};

/* Function that takes struct by value (copies the data) */
void print_point(struct Point p) {
    printf("Point(%d, %d)\n", p.x, p.y);
}

/* Function that takes struct by pointer (no copy, can modify) */
void move_point(struct Point *p, int dx, int dy) {
    p->x += dx;  /* Arrow operator for pointer->member */
    p->y += dy;
}

int main(void) {
    /* Initialize struct with designated initializers */
    struct Point origin = { .x = 0, .y = 0 };

    /* Initialize struct in order */
    struct Point corner = { 10, 20 };

    /* Declare then assign fields */
    struct Person student;
    strncpy(student.name, "Alice", sizeof(student.name) - 1);
    student.name[sizeof(student.name) - 1] = '\0';
    student.age = 20;
    student.height = 1.65;

    /* Using typedef struct */
    Token tok = { .text = "hello", .line = 1, .column = 5 };

    /* Nested struct initialization */
    struct Rectangle rect = {
        .top_left = { 0, 0 },
        .bottom_right = { 100, 50 }
    };

    printf("Origin: ");
    print_point(origin);

    printf("Corner before move: ");
    print_point(corner);

    move_point(&corner, 5, -3);  /* Pass address with & */
    printf("Corner after move(5, -3): ");
    print_point(corner);

    printf("\nStudent: %s, age %d, height %.2fm\n",
           student.name, student.age, student.height);

    printf("\nToken: '%s' at line %d, column %d\n",
           tok.text, tok.line, tok.column);

    printf("\nRectangle from (%d,%d) to (%d,%d)\n",
           rect.top_left.x, rect.top_left.y,
           rect.bottom_right.x, rect.bottom_right.y);

    return 0;
}

File: week01/rust/src/bin/03_structs.rs
//! 03_structs.rs - Struct definition with impl block
//!
//! Structs group related data together. Impl blocks add
//! methods and associated functions to structs.

/// A simple 2D point
#[derive(Debug, Clone, Copy)]  // Derive useful traits
struct Point {
    x: f64,
    y: f64,
}

/// Implementation block for Point
impl Point {
    /// Associated function (like a static method) - no self
    /// Called with Point::new(x, y)
    fn new(x: f64, y: f64) -> Point {
        Point { x, y }
    }

    /// Another associated function - creates origin point
    fn origin() -> Point {
        Point { x: 0.0, y: 0.0 }
    }

    /// Method taking &self (borrows, read-only)
    fn distance_from_origin(&self) -> f64 {
        (self.x * self.x + self.y * self.y).sqrt()
    }

    /// Method taking &mut self (borrows, can modify)
    fn translate(&mut self, dx: f64, dy: f64) {
        self.x += dx;
        self.y += dy;
    }
}

/// A rectangle defined by two corners
#[derive(Debug)]
struct Rectangle {
    top_left: Point,
    bottom_right: Point,
}

impl Rectangle {
    fn new(x1: f64, y1: f64, x2: f64, y2: f64) -> Rectangle {
        Rectangle {
            top_left: Point::new(x1, y1),
            bottom_right: Point::new(x2, y2),
        }
    }

    fn width(&self) -> f64 {
        (self.bottom_right.x - self.top_left.x).abs()
    }

    fn height(&self) -> f64 {
        (self.bottom_right.y - self.top_left.y).abs()
    }

    fn area(&self) -> f64 {
        self.width() * self.height()
    }

    fn contains(&self, point: &Point) -> bool {
        point.x >= self.top_left.x
            && point.x <= self.bottom_right.x
            && point.y >= self.top_left.y
            && point.y <= self.bottom_right.y
    }
}

/// Struct with String field (owned data)
#[derive(Debug)]
struct Person {
    name: String,
    age: u32,
}

impl Person {
    fn new(name: &str, age: u32) -> Person {
        Person {
            name: name.to_string(),
            age,  // Field init shorthand when variable name matches
        }
    }

    fn greet(&self) -> String {
        format!("Hello, I'm {} and I'm {} years old", self.name, self.age)
    }

    fn have_birthday(&mut self) {
        self.age += 1;
        println!("{} is now {} years old!", self.name, self.age);
    }
}

fn main() {
    println!("=== Creating Structs ===");

    // Using associated function (constructor)
    let p1 = Point::new(3.0, 4.0);
    let origin = Point::origin();

    // Direct initialization
    let p2 = Point { x: 1.0, y: 2.0 };

    println!("p1 = {:?}", p1);
    println!("origin = {:?}", origin);
    println!("p2 = {:?}", p2);

    println!("\n=== Calling Methods ===");
    println!("p1 distance from origin: {}", p1.distance_from_origin());

    // Mutable method requires mutable variable
    let mut p3 = Point::new(0.0, 0.0);
    println!("p3 before translate: {:?}", p3);
    p3.translate(5.0, 3.0);
    println!("p3 after translate(5, 3): {:?}", p3);

    println!("\n=== Nested Structs ===");
    let rect = Rectangle::new(0.0, 0.0, 10.0, 5.0);
    println!("Rectangle: {:?}", rect);
    println!("Width: {}, Height: {}", rect.width(), rect.height());
    println!("Area: {}", rect.area());

    let test_point = Point::new(5.0, 2.5);
    let outside_point = Point::new(15.0, 2.5);
    println!("{:?} inside rect: {}", test_point, rect.contains(&test_point));
    println!("{:?} inside rect: {}", outside_point, rect.contains(&outside_point));

    println!("\n=== Struct with String Field ===");
    let mut alice = Person::new("Alice", 30);
    println!("{:?}", alice);
    println!("{}", alice.greet());
    alice.have_birthday();

    println!("\n=== Struct Update Syntax ===");
    let p4 = Point { x: 100.0, ..p1 };  // Copy y from p1
    println!("p4 (x=100, y from p1): {:?}", p4);
}
